url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/core/QuartzSchedulerThread.java

```diff
@@ -0,0 +1,476 @@
+    this.quartzScheduler = quartzScheduler;
+    this.quartzSchedulerResources = quartzSchedulerResources;
+    this.setDaemon(setDaemon);
+    if (quartzSchedulerResources.isThreadsInheritInitializersClassLoadContext()) {
+      logger.info("QuartzSchedulerThread Inheriting ContextClassLoader of thread: " + Thread.currentThread().getName());
+      this.setContextClassLoader(Thread.currentThread().getContextClassLoader());
+    }
+
+    this.setPriority(threadPrio);
+
+    // start the underlying thread, but put this object into the 'paused'
+    // state
+    // so processing doesn't start yet...
+    paused = true;
+    halted = new AtomicBoolean(false);
+  }
+
+  /*
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Interface.
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+
+  private long getRandomizedIdleWaitTime() {
+
+    return idleWaitTime - random.nextInt(idleWaitVariablness);
+  }
+
+  /**
+   * <p>
+   * Signals the main processing loop to pause at the next possible point.
+   * </p>
+   */
+  public void togglePause(boolean pause) {
+
+    synchronized (sigLock) {
+      paused = pause;
+
+      if (paused) {
+        signalSchedulingChange(0);
+      } else {
+        sigLock.notifyAll();
+      }
+    }
+  }
+
+  /**
+   * <p>
+   * Signals the main processing loop to pause at the next possible point.
+   * </p>
+   */
+  public void halt() {
+
+    synchronized (sigLock) {
+      halted.set(true);
+
+      if (paused) {
+        sigLock.notifyAll();
+      } else {
+        signalSchedulingChange(0);
+      }
+    }
+  }
+
+  public boolean isPaused() {
+
+    return paused;
+  }
+
+  /**
+   * <p>
+   * Signals the main processing loop that a change in scheduling has been made - in order to interrupt any sleeping that may be occurring while
+   * waiting for the fire time to arrive.
+   * </p>
+   *
+   * @param candidateNewNextFireTime the time (in millis) when the newly scheduled trigger will fire. If this method is being called do to some other
+   *        even (rather than scheduling a trigger), the caller should pass zero (0).
+   */
+  void signalSchedulingChange(long candidateNewNextFireTime) {
+
+    synchronized (sigLock) {
+      signaled = true;
+      signaledNextFireTime = candidateNewNextFireTime;
+      sigLock.notifyAll();
+    }
+  }
+
+  private void clearSignaledSchedulingChange() {
+
+    synchronized (sigLock) {
+      signaled = false;
+      signaledNextFireTime = 0;
+    }
+  }
+
+  public boolean isScheduleChanged() {
+
+    synchronized (sigLock) {
+      return signaled;
+    }
+  }
+
+  public long getSignaledNextFireTime() {
+
+    synchronized (sigLock) {
+      return signaledNextFireTime;
+    }
+  }
+
+  /**
+   * <p>
+   * The main processing loop of the <code>QuartzSchedulerThread</code>.
+   * </p>
+   */
+  @Override
+  public void run() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.